### DISCUSSION TITLE
@storybook/angular writing stories for directives

### DISCUSSION THREAD

**Discussion by sir-captainmorgan21** [storybookjs/storybook#17491] [Help] [answered]
Has anyone been able to write stories for directives? Directives are a huge part of Angular. I've been able to follow some examples I have found elsewhere, but I cant get it to show my directive being used on a button in the code snippet under the preview like so

```
<button myButton>Click Me!</button>
```

Instead I get the dummy component that is being used from following this example... https://medium.com/@redin.gaetan/story-for-an-angular-directive-656885b44c79 like so

```
<dummy-component myinput='examle'></dummy-component>
```

**Comment by sir-captainmorgan21** [comment]
OK Update. That guys blog is out of date from the code that gets deployed to his live storybook example he links to. I was able to get it working with this:

```
import { CommonModule } from '@angular/common';
import { Story, Meta } from '@storybook/angular';
import { SAJButtonDirective } from './button.directive';

export default {
  title: 'SAJButtonDirective',
  component: SAJButtonDirective
} as Meta<SAJButtonDirective>;

const Template: Story<SAJButtonDirective> = (args: SAJButtonDirective) => ({
  props: args, 
  moduleMetadata: {
    declarations: [SAJButtonDirective], 
    imports: [CommonModule]
  },
  template: `<button sajButton [color]="color" [style]="style">Click</button>`
});


export const Primary = Template.bind({});
```

Only thing is the code snippet never gets updated from `<button sajButton [color]="color" [style]="style">Click</button>`. The story is interactive and updates live. Is a bug warranted for the code snippet not updating with the preview?

**Comment by t-kaufmann** [comment]
Oct. 2023 version (Storybook v7) as `Story` is now deprecated:

```
import type { Meta, StoryObj } from '@storybook/angular';
import { ButtonDirective } from './button.directive';

const meta: Meta<ButtonDirective> = {
  title: 'Components/Button',
  component: ButtonDirective,
  tags: ['autodocs'],
};

export default meta;
type Story = StoryObj<ButtonDirective>;

export const Default: Story = {
  render: (args) => ({
    props: args,
    template: `<button button-directive>Label</button>`,
  }),
};
```

**Reply by hard-rox** [reply]
That's made my day man. A big applause. 👏

**Comment by hard-rox** [comment, marked as answer]
With sb 7.6 this approach is also working.

```ts
const meta: Meta<ButtonDirective> = {
  component: ButtonDirective,
  title: 'directives/knButton',
  tags: ['autodocs'],
  render: (args) => ({
    props: args,
    template: `<button knButton>Hello Button</button>`
  }),
};
export default meta;
type Story = StoryObj<ButtonDirective>;

export const Default: Story = {
  play: async ({ canvasElement }) => {
    const canvas = within(canvasElement);
    expect(canvas.getByText(/pagination works!/gi)).toBeTruthy();
  },
};
```

**Reply by joevaugh4n** [reply]
Thank you, @hard-rox!

**Reply by aghwotu** [reply]
Thank you, @hard-rox . This was very helpful. How would we use multiple Directives for a story?

**Comment by RobbyRabbitman** [comment]
It's working for me except the 'autodocs' part: In the docs page it says: 'Args table with interactive controls couldn't be auto-generated'. When I convert the directive to a component it works, but thats dirty. Is it working for you?

**Reply by umutesen** [reply]
You need to ensure directives are included in the compodoc json. 

Update your storybook tsconfig to:

```
  "include": [
    "../../../packages/angular-components/**/*.directive.ts",
    "../../../packages/angular-components/**/*.component.ts",
//etc
  ]
```
